### ISSUE TITLE
Auto-grad during training

### ISSUE THREAD

**Issue by edgarriba** [tiny-dnn/tiny-dnn#539]
Could be nice to introduce a library to handle automatic differentiation during the training.

**Comment by bhack** [comment]
Want we close https://github.com/tiny-dnn/tiny-dnn/issues/223?

**Comment by soumith** [comment]
@edgarriba yes you could prob use it when it's fully fleshed out. here's the current ongoing development: https://github.com/colesbury/pytorch/tree/autograd/torch/csrc/autograd

**Comment by soumith** [comment]
alternatively, you can also add it via dynet which is C++ autograd library

**Comment by edgarriba** [comment]
@bhack maybe yes

@soumith but what's the status of this project? And are you going to release it as a standalone library and totally independent of Pytorch? We're just looking for potential projects to propose as ideas for this coming GSoC.

**Comment by soumith** [comment]
we dont have plans of releasing it as an independent project. it will be an independent library within the pytorch repo.

**Comment by bhack** [comment]
I think that is embedded with pytorch Tensor like [Dali](https://github.com/dali-ml/Dali) is embedded with mshadow. And we have our own header only Tensor.

**Comment by edgarriba** [comment]
@bhack ideally we should find something that it's "independent" of the Tensor library

**Comment by bhack** [comment]
I don't think that exist something similar. De facto Tensor is not standardized in c++xx and every framework has its own :)

**Comment by soumith** [comment]
see https://github.com/dmlc/mxnet/issues/4735 for more about standardized Tensor.

**Comment by bhack** [comment]
@soumith I've commented there about the Kronos effort to standardize Tensor api.

**Comment by tqchen** [comment]
depending on what you want to do. If you want symbolic automatic differentiation in computation graph, [NNVM](https://github.com/dmlc/nnvm) provide a compute invariant way to do so, and here is an example  project that uses the symbolic diff https://github.com/tqchen/tinyflow

**Comment by Randl** [comment]
https://github.com/BachiLi/had - here is AD lib with a review of other libs

**Comment by Randl** [comment]
http://www.autodiff.org/?module=Tools&language=C%2FC%2B%2B

**Comment by bhack** [comment]
Anybody seen https://zigasajovic.github.io/dCpp/?
